### PR TITLE
refactor(storage): improve retry error handling on resumable upload limit exceeded

### LINKED ISSUE #7346: How do I get more information about failing uploads?
Some of our servers crashed overnight when trying to upload data to GCS. The error was:

```
error: {
message: "Retry limit exceeded - """
name: "Error"
stack: "Error: Retry limit exceeded - ""
    at Upload.attemptDelayedRetry (/home/node/app/Backend/node_modules/@google-cloud/storage/build/cjs/src/resumable-upload.js:822:26)
    at Upload.onResponse (/home/node/app/Backend/node_modules/@google-cloud/storage/build/cjs/src/resumable-upload.js:786:18)
    at Upload.makeRequestStream (/home/node/app/Backend/node_modules/@google-cloud/storage/build/cjs/src/resumable-upload.js:772:40)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Upload.startUploading (/home/node/app/Backend/node_modules/@google-cloud/storage/build/cjs/src/resumable-upload.js:573:26)"
}
```

The actual reason why the uploads were failing is missing, though. Do I need to change something in the library's configuration to log that data? And what is the default retry limit value?

I am on version 7.14.0.

### PR DESCRIPTION
**Pull request by thiyaguk09**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/{{metadata['repo']['name']}}/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #7346

### CODE REVIEW

File: handwritten/storage/src/resumable-upload.ts [hunk 1]
```diff
@@ -1345,7 +1345,7 @@ export class Upload extends Writable {
       },
     };
     const res = await this.authClient.request(combinedReqOpts);
-    const successfulRequest = this.onResponse(res);
+    const successfulRequest = await this.onResponse(res);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/src/resumable-upload.ts]: removed

File: handwritten/storage/src/resumable-upload.ts [hunk 2]
```diff
@@ -1363,7 +1363,7 @@ export class Upload extends Writable {
         name: resp.statusText,
       })
     ) {
-      this.attemptDelayedRetry(resp);
+      await this.attemptDelayedRetry(resp);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/src/resumable-upload.ts]: removed

File: handwritten/storage/test/resumable-upload.ts [hunk 3]
```diff
@@ -2275,10 +2275,10 @@ describe('resumable-upload', () => {
...
       const RESP = {status: 500, data: 'error message from server'};
 
-      it('should increase the retry count if less than limit', () => {
+      it('should increase the retry count if less than limit', async () => {
         up.getRetryDelay = () => 1;
         assert.strictEqual(up.numRetries, 0);
-        assert.strictEqual(up.onResponse(RESP), false);
+        assert.strictEqual(await up.onResponse(RESP), false);
         assert.strictEqual(up.numRetries, 1);
       });
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: removed

File: handwritten/storage/test/resumable-upload.ts [hunk 4]
```diff
@@ -2321,19 +2323,19 @@ describe('resumable-upload', () => {
             assert(delay <= maxTime);
 
             // make it keep retrying until the limit is reached
-            up.onResponse(RESP);
+            void up.onResponse(RESP);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: removed

File: handwritten/storage/test/resumable-upload.ts [hunk 5]
```diff
@@ -2348,23 +2350,22 @@ describe('resumable-upload', () => {
           assert.strictEqual(resp, RESP);
           done();
         });
-        up.onResponse(RESP);
+        void up.onResponse(RESP);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: removed

File: handwritten/storage/test/resumable-upload.ts [hunk 6]
```diff
@@ -2348,23 +2350,22 @@ describe('resumable-upload', () => {
...
       });
 
-      it('should return true', () => {
+      it('should return true', async () => {
         up.getRetryDelay = () => 1;
-        assert.strictEqual(up.onResponse(RESP), true);
+        assert.strictEqual(await up.onResponse(RESP), true);
       });
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: reverted

File: handwritten/storage/test/resumable-upload.ts [hunk 7]
```diff
@@ -2348,23 +2350,22 @@ describe('resumable-upload', () => {
...
         up.getRetryDelay = () => 1;
         const RESP = {status: 1000};
         const customHandlerFunction = (err: ApiError) => {
           return err.code === 1000;
         };
         up.retryOptions.retryableErrorFn = customHandlerFunction;
-
-        assert.strictEqual(up.onResponse(RESP), false);
+        assert.strictEqual(await up.onResponse(RESP), false);
       });
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: reverted

File: handwritten/storage/test/resumable-upload.ts [hunk 8]
```diff
@@ -2321,19 +2323,19 @@ describe('resumable-upload', () => {
...
             assert.strictEqual(
               err.message,
-              `Retry limit exceeded - ${JSON.stringify(RESP.data)}`
+              `Retry limit exceeded - status: 500 - error message from server`,
             );
             done();
           });
 
-          up.onResponse(RESP);
+          void up.onResponse(RESP);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: reverted

File: handwritten/storage/test/resumable-upload.ts [hunk 9]
```diff
@@ -2287,15 +2287,17 @@ describe('resumable-upload', () => {
...
-        up.onResponse(RESP);
-        up.onResponse(RESP);
-        up.onResponse(RESP);
-        up.onResponse(RESP);
+        (async () => {
+          await up.onResponse(RESP);
+          await up.onResponse(RESP);
+          await up.onResponse(RESP);
+          await up.onResponse(RESP);
+        })().catch(done);
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: reverted

File: handwritten/storage/src/resumable-upload.ts [hunk 10]
```diff
@@ -1354,7 +1354,7 @@ export class Upload extends Writable {
   /**
    * @return {bool} is the request good?
    */
-  private onResponse(resp: GaxiosResponse) {
+  private async onResponse(resp: GaxiosResponse) {
```
[the remark this answers is not in the document]
Comment (thiyaguk09) [reply handwritten/storage/src/resumable-upload.ts]: removed

**Review by Dhriti07 — COMMENTED** [review]
It looks like you reverted the async changes to onResponse in your recent commit, but the PR title still mentions it. Is that something this pr aims to do ?

File: handwritten/storage/test/resumable-upload.ts [hunk 11]
```diff
@@ -2490,6 +2489,80 @@ describe('resumable-upload', () => {
...
+        done();
+      });
+
+      up.attemptDelayedRetry({
+        status: NaN,
+        data: customError,
+      });
+    });
+
+    it('should include correct details for GaxiosErrors with empty/missing response bodies', done => {
```
Comment (Dhriti07) [inline handwritten/storage/test/resumable-upload.ts:2532]: The tests for GaxiosError cover the edge case where the response body is empty.  It would be great to add one more test for a populated error response

File: handwritten/storage/test/resumable-upload.ts [hunk 12]
```diff
@@ -2490,6 +2489,80 @@ describe('resumable-upload', () => {
...
+          status: 429,
+          statusText: 'Too Many Requests',
+          data: '',
+          config: {},
+          headers: {},
+        } as any
+      );
+
+      up.on('error', (err: Error) => {
+        // Let's check for the presence of key details rather than an exact string if we are unsure of exact GaxiosError fields,
```
Comment (Dhriti07) [inline handwritten/storage/test/resumable-upload.ts]: Nit: This comment is a bit conversational. To align with readability guidelines, could we simplify it to directly explain the why?

Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts]: Done.

File: handwritten/storage/test/resumable-upload.ts (hunk 11 above)
Comment (thiyaguk09) [reply handwritten/storage/test/resumable-upload.ts:2532]: Done. Added test case.

**Comment by thiyaguk09** [comment]
> It looks like you reverted the async changes to onResponse in your recent commit, but the PR title still mentions it. Is that something this pr aims to do ?

I updated the PR title to reflect this change.

**Review by Dhriti07 — APPROVED** [review]
